zquery/shared/src/main/scala/zio/zquery/Continue.scala

```diff
@@ -0,0 +1,140 @@
+    self match {
+      case Get(io)       => get(io.mapError(f))
+      case Effect(query) => effect(query.mapError(f))
+    }
+
+  /**
+   * Effectually maps over the success type of this continuation.
+   */
+  def mapM[R1 <: R, E1 >: E, B](f: A => ZQuery[R1, E1, B]): Continue[R1, E1, B] =
+    self match {
+      case Get(io)       => effect(ZQuery.fromEffect(io).flatMap(f))
+      case Effect(query) => effect(query.flatMap(f))
+    }
+
+  /**
+   * Purely contramaps over the environment type of this continuation.
+   */
+  def provideSome[R0](f: Described[R0 => R])(implicit ev: NeedsEnv[R]): Continue[R0, E, A] =
+    self match {
+      case Get(io)       => get(io)
+      case Effect(query) => effect(query.provideSome(f))
+    }
+
+  /**
+   * Runs this continuation..
+   */
+  def runCache(cache: Cache): ZIO[R, E, A] =
+    self match {
+      case Get(io)       => io
+      case Effect(query) => query.runCache(cache)
+    }
+
+  /**
+   * Combines this continuation with that continuation using the specified
+   * function, in sequence.
+   */
+  def zipWith[R1 <: R, E1 >: E, B, C](that: Continue[R1, E1, B])(f: (A, B) => C): Continue[R1, E1, C] =
+    (self, that) match {
+      case (Get(l), Get(r))       => get(l.zipWith(r)(f))
+      case (Get(l), Effect(r))    => effect(ZQuery.fromEffect(l).zipWith(r)(f))
+      case (Effect(l), Get(r))    => effect(l.zipWith(ZQuery.fromEffect(r))(f))
+      case (Effect(l), Effect(r)) => effect(l.zipWith(r)(f))
+    }
+
+  /**
+   * Combines this continuation with that continuation using the specified
+   * function, in parallel.
+   */
+  def zipWithPar[R1 <: R, E1 >: E, B, C](that: Continue[R1, E1, B])(f: (A, B) => C): Continue[R1, E1, C] =
+    (self, that) match {
+      case (Get(l), Get(r))       => get(l.zipWithPar(r)(f))
+      case (Get(l), Effect(r))    => effect(ZQuery.fromEffect(l).zipWithPar(r)(f))
+      case (Effect(l), Get(r))    => effect(l.zipWithPar(ZQuery.fromEffect(r))(f))
+      case (Effect(l), Effect(r)) => effect(l.zipWithPar(r)(f))
+    }
+
+}
+
+private[zquery] object Continue {
+
+  /**
+   * Constructs a continuation from a request, a data source, and a `Ref` that
+   * will contain the result of the request when it is executed.
+   */
+  def apply[R, E, A, B](request: A, dataSource: DataSource[R, A], ref: Ref[Option[Either[E, B]]])(
+    implicit ev: A <:< Request[E, B]
+  ): Continue[R, E, B] =
+    Continue.get {
+      ref.get.flatMap {
+        case None    => IO.die(QueryFailure(dataSource, request))
+        case Some(b) => IO.fromEither(b)
+      }
+    }
+
+  /**
+   * Constructs a continuation that may perform arbitrary effects.
+   */
+  def effect[R, E, A](query: ZQuery[R, E, A]): Continue[R, E, A] =
+    Effect(query)
+
+  /**
+   * Constructs a continuation that merely gets the result of a blocked request
+   * (potentially transforming it with pure functions).
+   */
+  def get[E, A](io: IO[E, A]): Continue[Any, E, A] =
```

Comment:
`fromEffect`???

Reply:
Also, is it possible to generalize to include `R` here?

Reply:
I'm worried this makes it less clear, though I also think this indicates I haven't made it clear enough to begin with. 😃 

The problem we're trying to solve here is to distinguish between:

```scala
request1.map(f) *> request2
// okay to pipeline requests and perform `f` after because it is a pure function

request.flatMap(_ => g) *> request2
// not safe to pipeline requests because `g` must be performed before `request2`
```

So we want to say that the effect in `Get` can't do anything more than get the result of the request and transform it with pure functions. That way in `ZQuery#zipWith` we can introspect on the continuation to determine whether we can pipeline.

So I worry that `fromEffect` is a confusing name because we don't want to put arbitrary effects in here. That's why I didn't allow an `R` type. If anything to enforce it through the types we might want to do something like:

```scala
sealed trait Continue[-R, +E, +A]

final case class Effect[-R, +E, +A](query: ZQuery[R, E, A]) extends Continue[R, E, A]
final case class Get[+E, +A](ref: ZRef[Any, E, Nothing, A) extends Continue[Any, E, A]
```

Then we would be enforcing that if you just want to purely transform the value you can do that within `Get` by transforming the `ZRef` but if you want to perform any other type of effects you need to return an `Effect`.

Reply:
So the provided `io` effect is supposed to _not_ do anything except get from a ref and transform it? If so, maybe we should store a pure function instead of an IO effect?

Reply:
Yes. The problem is we need somewhere to put the result of the request when it is completed. So we could have a `Ref` plus a function or just a `ZRef`.

Reply:
Also we want to be stack safe if for example someone is doing a long chain of maps on the continuation.

Reply:
Do you want to pair on this? Or we can merge as-is, and think about it later in a follow-up.

Reply:
That would be great!